lib/models/event-timeline.js

```diff
@@ -0,0 +1,260 @@
+ */
+
+var RoomState = require("./room-state");
+
+/**
+ * Construct a new EventTimeline
+ *
+ * <p>An EventTimeline represents a contiguous sequence of events in a room.
+ *
+ * <p>As well as keeping track of the events themselves, it stores the state of
+ * the room at the beginning and end of the timeline, and pagination tokens for
+ * going backwards and forwards in the timeline.
+ *
+ * <p>In order that clients can meaningfully maintain an index into a timeline,
+ * the EventTimeline object tracks a 'baseIndex'. This starts at zero, but is
+ * incremented when events are prepended to the timeline. The index of an event
+ * relative to baseIndex therefore remains constant.
+ *
+ * <p>Once a timeline joins up with its neighbour, they are linked together into a
+ * doubly-linked list.
+ *
+ * @param {string} roomId    the ID of the room where this timeline came from
+ * @constructor
+ */
+function EventTimeline(roomId) {
+    this._roomId = roomId;
+    this._events = [];
+    this._baseIndex = -1;
+    this._startState = new RoomState(roomId);
+    this._endState = new RoomState(roomId);
+
+    this._prevTimeline = null;
+    this._nextTimeline = null;
+
+    // this is used by client.js
+    this._paginationRequests = {'b': null, 'f': null};
+}
+
+/**
+ * Initialise the start and end state with the given events
+ *
+ * <p>This can only be called before any events are added.
+ *
+ * @param {MatrixEvent[]} stateEvents list of state events to initialise the
+ * state with.
+ * @throws {Error} if an attempt is made to call this after addEvent is called.
+ */
+EventTimeline.prototype.initialiseState = function(stateEvents) {
+    if (this._events.length > 0) {
+        throw new Error("Cannot initialise state after events are added");
+    }
+
+    // do we need to copy here? sync thinks we do but I can't see why
+    this._startState.setStateEvents(stateEvents);
+    this._endState.setStateEvents(stateEvents);
+};
+
+/**
+ * Get the ID of the room for this timeline
+ * @return {string} room ID
+ */
+EventTimeline.prototype.getRoomId = function() {
+    return this._roomId;
+};
+
+/**
+ * Get the base index.
+ *
+ * <p>This is an index which is incremented when events are prepended to the
+ * timeline. An individual event therefore stays at the same index in the array
+ * relative to the base index.
+ *
+ * @return {number}
+ */
+EventTimeline.prototype.getBaseIndex = function() {
+    return this._baseIndex;
+};
+
+/**
+ * Get the list of events in this context
+ *
+ * @return {MatrixEvent[]} An array of MatrixEvents
+ */
+EventTimeline.prototype.getEvents = function() {
+    return this._events;
+};
+
+/**
+ * Get the room state at the start/end of the timeline
+ *
+ * @param {boolean} start true to get the state at the start of the timeline;
+ * false to get the state at the end of the timeline.
+ * @return {RoomState} state at the start/end of the timeline
+ */
+EventTimeline.prototype.getState = function(start) {
```

Comment:
What I'm going to say is going to be a pain but I think it will be for the best.

I don't like the boolean switching for start/end and would prefer `getStartState()` and `getEndState()`. Why? Simply because when you're reading code which uses this (for me it was when I was reading the tests) it is not immediately obvious what `getState(true)` refers to. I then need to remember the line I'm on, hunt the method definition and check what the `boolean` means. This makes me sad. Legibility of code can be improved by using more descriptive function names IMO. I appreciate this is nitty and a PITA :(